packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx

```diff
@@ -382,84 +457,349 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        optionsRef.current = Array.from(items) as HTMLLIElement[]; // need this cast to make TS happy :(
+      }
+    }, [menuRef, optionsRef, props.children]);
+
+    const refs = useMergedRefs(menuRef, forwardedRef);
+
+    return (
+      <Menu
+        id={`${id}-list`}
+        style={React.useMemo(
+          () => ({
+            minWidth,
+            maxWidth: `min(${minWidth * 2}px, 90vw)`,
+            maxHeight: 300,
+            ...style,
+          }),
+          [minWidth, style],
+        )}
+        setFocus={false}
+        role='listbox'
+        ref={refs}
+        className={cx('iui-scroll', className)}
+        {...rest}
+      />
+    );
+  },
+);
+ComboBoxMenu.displayName = 'ComboBoxMenu';
+
+const ComboBoxMenuItem = React.memo(
+  React.forwardRef(
+    (
+      props: MenuItemProps & { index: number },
+      forwardedRef: React.RefObject<HTMLLIElement>,
+    ) => {
+      const {
+        children,
+        isSelected,
+        disabled,
+        value,
+        onClick,
+        sublabel,
+        size = !!sublabel ? 'large' : 'default',
+        icon,
+        badge,
+        className,
+        role = 'menuitem',
+        index,
+        ...rest
+      } = props;
+
+      const { focusedIndex } = useSafeContext(ComboBoxStateContext);
+
+      const focusRef = (el: HTMLLIElement | null) => {
+        // will need to check for virtualization here too
+        if (focusedIndex === index) {
+          el?.scrollIntoView({ block: 'nearest' });
+        }
+      };
+
+      const refs = useMergedRefs(forwardedRef, focusRef);
+
+      return (
+        <li
+          className={cx(
+            'iui-menu-item',
+            {
+              'iui-large': size === 'large',
+              'iui-active': isSelected,
+              'iui-disabled': disabled,
+              'iui-focused': focusedIndex === index,
+            },
+            className,
+          )}
+          ref={refs}
+          onClick={() => !disabled && onClick?.(value)}
+          role={role}
+          tabIndex={disabled || role === 'presentation' ? undefined : -1}
+          aria-selected={isSelected}
+          aria-disabled={disabled}
+          data-iui-index={index}
+          {...rest}
         >
-          <SvgCaretDownSmall aria-hidden />
-        </span>
-      </div>
-    </InputContainer>
-  );
-};
+          {icon &&
+            React.cloneElement(icon, {
+              className: cx(icon.props.className, 'iui-icon'),
+            })}
+          <span className='iui-content'>
+            <div className='iui-menu-label'>{children}</div>
+            {sublabel && <div className='iui-menu-description'>{sublabel}</div>}
+          </span>
+          {badge &&
+            React.cloneElement(badge, {
+              className: cx(badge.props.className, 'iui-icon'),
+            })}
+        </li>
+      );
+    },
+  ),
+);
+ComboBoxMenuItem.displayName = 'ComboBoxMenuItem';
+
+ComboBox.InputContainer = ComboBoxInputContainer;
```

Comment:
Maybe have subcomponents in different files like in other our components. Because it is hard to view ~1k lines of code.

Reply:
i think @veekeys will disagree

Reply:
Carousel we  did separate in files. So I guess lets be consistent..?

Reply:
I agree to put subcompontents in different files

Reply:
I've split everything into separate files 👍

Reply:
But you do not export them. On purpose just as internal use?

Reply:
Yup, lets keep internal for now. We can expose a better API with a better custom example in a future PR